### DISCUSSION TITLE
HTTP에서 자주 사용되는 GET과 POST를 비교 설명해주세요

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#171] [Network]
.

**Comment by J-I-H-O** [comment]
GET 메서드와 POST 메서드는 다음과 같은 차이점이 있습니다.

- **사용 목적**
    - GET: 서버로부터 정보를 요청할 때 사용되는 HTTP 메서드 입니다.
    - POST: 정보를 생성/업데이트 하기 위해 서버에 데이터를 보내는 HTTP 메서드 입니다.
- **캐시 여부**
    - GET: 불필요한 요청을 제한하기 위해 요청이 캐시될 수 있습니다. 브라우저나 프록시 서버는 GET 요청의 응답을 캐싱하여 동일한 요청에 대해 서버에 재요청하지 않고, 캐시된 데이터를 사용할 수 있습니다.
    - POST: POST 요청은 기본적으로 캐싱되지 않습니다. 서버는 POST 요청을 받을 때 마다 새로운 처리를 수행합니다.
- **body 유무 여부**
    - GET: 요청 데이터를 쿼리 파라미터로 보내기 때문에 HTTP 메시지에 body가 없습니다.
    이처럼 GET 요청은 데이터가 URL에 노출되고, 브라우저 히스토리에 기록되기 때문에 민감한 정보를 전송하는데 적합하지 않습니다.
    또 URL은 길이에 제한이 있기 때문에 GET 요청으로 전송할 수 있는 데이터의 크기는 제한적입니다.
    - POST: body에 요청 데이터를 담아 보내기 때문에 HTTP 메시지에 body가 존재합니다. 
    이같은 특성으로 인해 URL에 민감한 정보가 노출되지 않고, 브라우저 히스토리에 기록되지 않습니다. 하지만 여전히 HTTPS를 사용해 데이터를 암호화 하는 것이 좋습니다.
    요청 body에 데이터를 포함하므로 전송할 수 있는 데이터의 크기에 제한이 거의 없습니다.
- **멱등성**
    - GET 메서드는 멱등이고, POST 요청은 멱등이 아닙니다.
    이때 **멱등**이란, **연산을 여러 번 적용 하더라도 결과가 항상 달라지지 않는 성질**을 뜻합니다. 
    GET은 리소스를 단순 조회하기 때문에 여러번 수행하더라도 응답이 똑같은 반면, POST는 리소스를 생성하거나 변경할 수 있기 때문에 멱등이 아닙니다.

https://www.w3schools.com/tags/ref_httpmethods.asp

**Comment by le2sky** [comment]
HTTP GET 메서드는 리소스 조회에 사용됩니다.
서버에 전달하고 싶은 데이터는 쿼리 파라미터를 이용해 전달합니다.
메시지 바디를 사용해서 데이터를 전달할 수도 있지만, 지원하지 않는 곳이 많아서 권장되지는 않습니다.

반면 HTTP POST 메서드는 요청 데이터를 처리하는 경우에 사용됩니다.
메시지 바디를 이용해 서버로 요청 데이터를 전달합니다. 
요청 데이터를 받은 서버는 전달된 데이터로 신규 리소스를 등록하거나, 프로세스를 처리합니다.

또한 GET은 여러번 호출해도 같은 결과가 조회되니 멱등한 반면,
POST는 여러번 호출하면 새로운 리소스가 계속해서 생성될 수 있으니 멱등하지 않습니다.(스펙에 대한 이야기)
